src/eg/src/guardian_proof.rs

```diff
@@ -0,0 +1,276 @@
+        v.extend_from_slice(to_be_bytes_left_pad(capital_k_i_j, 512).as_slice());
+        v.extend_from_slice(to_be_bytes_left_pad(h_i_j, 512).as_slice());
+        let c = eg_h(&h_p, &v);
+        //The challenge is not reduced modulo q (cf. Section 5.4)
+        PossessionProofChallenge(BigUint::from_bytes_be(c.0.as_slice()))
+    }
+
+    /// This function computes coefficient NIZK proof as specified in Section 3.2.2
+    /// The arguments are
+    /// - csprng - secure randomness generator
+    /// - election_parameters - the election parameters
+    /// - h_p - the parameter base hash
+    /// - secret_key - the guardian `GuardianSecretKey`
+    pub fn new(
+        csprng: &mut Csprng,
+        election_parameters: &ElectionParameters,
+        h_p: HValue,
+        secret_key: &GuardianSecretKey,
+    ) -> Self {
+        let fixed_parameters = &election_parameters.fixed_parameters;
+        let varying_parameters = &election_parameters.varying_parameters;
+        let k = varying_parameters.k;
+
+        // Compute commit message
+        let u_vec = (0..k.get_one_based_u32())
+            .map(|_j| csprng.next_biguint_lt(fixed_parameters.q.borrow()))
+            .collect::<Vec<BigUint>>();
+        let h_vec = u_vec
+            .iter()
+            .map(|u_i| fixed_parameters.g.modpow(u_i, fixed_parameters.p.as_ref()))
+            .collect::<Vec<BigUint>>();
+        // Compute challenge vector
+        let c_vec = (0..k.get_one_based_u32())
+            .map(|j| {
+                Self::challenge(
+                    h_p,
+                    secret_key.i.get_one_based_u32(),
+                    j,
+                    &secret_key.coefficient_commitments.0[j as usize].0,
+                    &h_vec[j as usize],
+                )
+            })
+            .collect::<Vec<PossessionProofChallenge>>();
+        // Compute response vector
+        let v_vec = (0..k.get_one_based_usize())
+            .map(|j| {
+                PossessionProofResponse(fixed_parameters.q.subtract_group_elem(
+                    &u_vec[j],
+                    &(&c_vec[j].0 * &secret_key.secret_coefficients.0[j].0),
+                ))
+            })
+            .collect::<Vec<PossessionProofResponse>>();
+        GuardianProof {
+            i: secret_key.i,
+            c: c_vec,
+            v: v_vec,
+        }
+    }
+
+    /// This function verifies a `GuardianProof` with respect to a given `GuardianPublicKeyInfo`
+    /// This is corresponds to Verification 2 "Guardian public-key validation"
+    /// The arguments are
+    /// - self - the `GuardianProof`
+    /// - election_parameters - the election parameters
+    /// - h_p - the parameter base hash
+    /// - public_key_info - the `GuardianPublicKeyInfo`
+    pub fn validate(
+        &self,
+        election_parameters: &ElectionParameters,
+        h_p: HValue,
+        public_key_info: &dyn GuardianPublicKeyInfo,
+    ) -> Result<()> {
```

Comment:
API-wise, it'd be better to define a proper error type here instead of using `anyhow::Error`.

Something like

```
#[derive(Debug, thiserror::Error)]
enum ValidationError {
   #[error("The guardian indices of the proof {proof} and the public key {public_key} must match.
   IndexMismatch {
       proof: u32,
       public_key: u32
   }
}
```

Reply:
Added proper errors.